### PR TITLE
Use Basic auth for credentials instead of query params

### PR DESCRIPTION
**Pull request by enqueue**
Fix for #522

### CODE REVIEW

**Comment by zhicwu** [comment]
Thank you @enqueue! Will it work when `httpAuthorization` is set?

**Comment by enqueue** [comment]
> Thank you @enqueue! Will it work when `httpAuthorization` is set?

Good point! I will check it.

File: src/main/java/ru/yandex/clickhouse/util/ClickHouseHttpClientBuilder.java [hunk 1]
```diff
@@ -209,4 +240,31 @@ private KeyStore getKeyStore()
...
+
+    private static HttpHost getTargetHost(ClickHouseProperties props) {
+        return new HttpHost(
+            props.getHost(),
+            props.getPort(),
+            props.getSsl() ? "https" : "http");
+    }
+
+    private static boolean isConfigurationValidForAuth(ClickHouseProperties props) {
+        return props.getHost() != null
```
Comment (zhicwu) [inline src/main/java/ru/yandex/clickhouse/util/ClickHouseHttpClientBuilder.java:266]: I'm not sure why we need to check host here - it's always `null` in `BalancedClickhouseDataSource`. Will this cause authentication issue when connecting to a cluster?

File: src/test/java/ru/yandex/clickhouse/util/ClickHouseHttpClientBuilderTest.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+            "basic");
+    }
+
+
+    @Test(dataProvider = "authUserPassword")
+    public void testHttpAuthParametersCombination(String authorization, String user,
+        String password, String expectedAuthHeader) throws Exception
+    {
+        ClickHouseProperties props = new ClickHouseProperties();
+        props.setHost("localhost");
```
Comment (zhicwu) [inline src/test/java/ru/yandex/clickhouse/util/ClickHouseHttpClientBuilderTest.java:106]: Could you add one more test when host is null?

File: src/main/java/ru/yandex/clickhouse/ClickHouseStatementImpl.java [hunk 3]
```diff
@@ -15,35 +31,33 @@
...
-import java.util.*;
-
 
 public class ClickHouseStatementImpl implements ClickHouseStatement {
 
     private static final Logger log = LoggerFactory.getLogger(ClickHouseStatementImpl.class);
 
     private final CloseableHttpClient client;
 
+    private final HttpClientContext httpContext;
```
Comment (zhicwu) [inline src/main/java/ru/yandex/clickhouse/ClickHouseStatementImpl.java:59]: Why not make it protected and create the context in constructor?

**Comment by zhicwu** [comment]
Thank you @enqueue for making this pull request. I think we're close to get it merged. Mean time, can we remove HttpContext parameter from Statement and PreparedStatement constructors? Instead, we can create http context inside the constructor. This is because HttpContext is tied to http, so when we add support for more protocols, we'll have to remove it anyway because we only need to pass a client(e.g. HttpClient, NativeClient or GrpcClient) for execution.

Let me know if you have different opinion and we can discuss.

File: src/main/java/ru/yandex/clickhouse/ClickHouseStatementImpl.java (hunk 3 above)
Comment (enqueue) [reply src/main/java/ru/yandex/clickhouse/ClickHouseStatementImpl.java:59]: This is analogous to the client. This class does not know how to construct these, but gets ready to use "connection stuff".

File: src/test/java/ru/yandex/clickhouse/util/ClickHouseHttpClientBuilderTest.java (hunk 2 above)
Comment (enqueue) [reply src/test/java/ru/yandex/clickhouse/util/ClickHouseHttpClientBuilderTest.java:106]: Sure

File: src/main/java/ru/yandex/clickhouse/util/ClickHouseHttpClientBuilder.java (hunk 1 above)
Comment (enqueue) [reply src/main/java/ru/yandex/clickhouse/util/ClickHouseHttpClientBuilder.java:266]: Good point. I had not thought about this. Will try to find out how the `BalancedClickhouseDataSource` works.

Comment (enqueue) [reply src/main/java/ru/yandex/clickhouse/util/ClickHouseHttpClientBuilder.java:266]: Added some tests for common scenarios to `BalancedClickhouseDataSourceTest`, do you have any others you can think of? The host is set in 'ClickhouseJdbcUrlParser.parseClickhouseUrl` from the URL, so we hopefully do not have to change the implementation of `BalancedClickhouseDataSource`

File: src/main/java/ru/yandex/clickhouse/ClickHouseStatementImpl.java (hunk 3 above)
Comment (enqueue) [reply src/main/java/ru/yandex/clickhouse/ClickHouseStatementImpl.java:60]: Well, I would say that the statement should simply get everything it needs to do its work. It does not need to know how to create connection context, but this is my subjective judgement. I can understand that we want to try and keep the number of ties with Apache HTTP Client under control. I am going to follow your suggestion.

**Comment by enqueue** [comment]
Hi @zhicwu  I updated the PR to take into account your suggestions. I need your help getting the build to green though. The maven output says something about `ClickHouseResponseSummary` missing, but there isn't any code referring to this class. Running mvn verify locally works. What am I doing wrong? Thank you very much in advance.

**Comment by zhicwu** [comment]
Thank you @enqueue. ResponseSummary was merged from PR #498. Have you merged all changes from upstream/develop? I'm out for launch but I'll look into this when I get back to home.

File: src/main/java/ru/yandex/clickhouse/ClickHouseStatementImpl.java (hunk 3 above)
Comment (zhicwu) [reply src/main/java/ru/yandex/clickhouse/ClickHouseStatementImpl.java:57]: I see your point. Perhaps it's too early to consider refactoring. I'm fine to keep HttpContext argument now, but I think we may use a new interface like ClickHouseClient to replace both CloseableHtppClient and HttpContext, in order to hide protocol implementation details.

File: src/test/java/ru/yandex/clickhouse/util/ClickHouseHttpClientBuilderTest.java (hunk 2 above)
Comment (zhicwu) [reply src/test/java/ru/yandex/clickhouse/util/ClickHouseHttpClientBuilderTest.java:106]: Thought host is null when using BalancedDataSource but I'll double check :p

File: src/main/java/ru/yandex/clickhouse/util/ClickHouseHttpClientBuilder.java (hunk 1 above)
Comment (zhicwu) [reply src/main/java/ru/yandex/clickhouse/util/ClickHouseHttpClientBuilder.java:266]: I see. Thanks for the detailed explanation.  I don't have anything to add then.
